### DISCUSSION TITLE
How to write AST selector to match only object key (and not also object value) ?

### DISCUSSION THREAD

**Discussion by lukasMega** [eslint/eslint#18320] [Q&A] [answered]
How to write AST selector to match only object key (and not also object value) ?

my rule (partially working) is:
 
- ```javascript
  eslint no-restricted-syntax: ["error", 
      {
          selector: 'Property[computed=true] > MemberExpression, Property[computed=true] > Literal',
          message: 'computed object properties are not allowed!',     
      }
  ]
  ```



here is [example](https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tcmVzdHJpY3RlZC1zeW50YXg6IFtcImVycm9yXCIsIFxuICAgIHtcbiAgICAgICAgc2VsZWN0b3I6ICdQcm9wZXJ0eVtjb21wdXRlZD10cnVlXSA+IE1lbWJlckV4cHJlc3Npb24sIFByb3BlcnR5W2NvbXB1dGVkPXRydWVdID4gTGl0ZXJhbCcsXG4gICAgICAgIG1lc3NhZ2U6ICdjb21wdXRlZCBvYmplY3QgcHJvcGVydGllcyBhcmUgbm90IGFsbG93ZWQhJywgICAgIFxuICAgIH1cbl0qL1xuXG5leGFtcGxlW1QuZXJyXSA9IHsgW1Muc3RvcHBlZF06IFMucnVubmluZyB9OyBcbi8qICAgICAgICAgICAgICAgICB+fn5+fn5+fn5+fiBcbiAgICAgICAgICAgICAgICAgICAgICAgIOKGqiDinZfvuI8gbWF0Y2ggb25seSBvYmplY3Qga2V5IGluIGNhc2UgaXQncyBjb21wdXRlZCBcbiovXG5leGFtcGxlW1Qub2trXSA9IHsgbmFtZU9rOiBTLnJ1bm5pbmcgfTsgLy8g4pyFIGhlcmUgaXMgYWxsIG9rLCBubyBtYXRjaFxuXG5jb25zdCBtID0gJCh7XG4gICAgW1MucmVhZHldOiB7ICAgICAgICAgICAgICAgICAgICAgICAvLyDinIUgaGVyZSBpcyBjb3JyZWN0bHkgbWF0Y2hlZCBvbmx5IG9iamVjdCBrZXkgKGNvbXB1dGVkKVxuICAgICAgICBhY3Rpb25zOiB7IG9uRW50ZXIsIG9uRXhpdCB9LFxuICAgICAgICB0cmFuc2l0aW9uczogeyBzdGFydCB9LFxuICAgIH0sXG4gICAgWydub3RPayddOiBub3RPa1ZhbHVlLCAgICAgICAgICAgIC8vICDinIUgaGVyZSBpcyBjb3JyZWN0bHkgbWF0Y2hlZCBvbmx5IG9iamVjdCBrZXkgKGNvbXB1dGVkKVxuICAgIG5hbWVPazogeyBzdHVmZiB9LFxufSk7Iiwib3B0aW9ucyI6eyJydWxlcyI6e30sImxhbmd1YWdlT3B0aW9ucyI6eyJzb3VyY2VUeXBlIjoic2NyaXB0IiwicGFyc2VyT3B0aW9ucyI6eyJlY21hRmVhdHVyZXMiOnt9fSwiZWNtYVZlcnNpb24iOjIwMTZ9fX0=):

```javascript
/*eslint no-restricted-syntax: ["error", 
    {
        selector: 'Property[computed=true]:has(MemberExpression) > MemberExpression',
        message: 'compute object properties are not allowed!',     
    }
]*/

example[T.err] = { [S.stopped]: S.running }; 
/*                 ~~~~~~~~~~~ 
                        ↪ ❗️ match only object key in case it's computed 
*/
example[T.okk] = { nameOk: S.running }; // ✅ here is all ok, no match

const m = $({
    [S.ready]: {                       // ✅ here is correctly matched only object key (computed)
        actions: { onEnter, onExit },
        transitions: { start },
    },
    ['notOk']: notOkValue,
    nameOk: { stuff },
});
```

My attempts:
- `Property[computed=true]:has(MemberExpression) > MemberExpression` -  is partially working
- `Property[computed=true] > MemberExpression` - the same

I tried to use **`:first-child`**:
- `Property[computed=true] > MemberExpression:first-child` - not working as I expected

I'm using also https://astexplorer.net/ to see generated tree from code.

Thank you in advance for help.

**Comment by lukasMega** [comment, marked as answer]
Ok, I found solution finally.
I hope that this information will be beneficial to those who encounter it in the future.

#### ✅ SOLUTION: 

- ### `ObjectExpression> Property[computed=true] > MemberExpression.key`
  - it's more specific

- ### `Property[computed=true] > :expression.key`
  - more generic probably, definitely shorter.

---

## ✅ SOLUTION (final):

//update:
Thank you @fisker, it's even shorter and working as expected:
- ## `Property[computed=true] > .key`

BTW this is nice place to try AST selector with ESLint, however it's for Typescript:  [https://typescript-eslint.io/play/...](https://typescript-eslint.io/play/#ts=4.6.2&showAST=es&fileType=.js&esQuery=N4IgZglgNgLgpgJxALhCAvkA&code=PQKgpgzgNglgdgFwARwPYFoBOkGZgYwTABN0IBPRAQwA8AuJAbQCIxNNVNmAaJAWABQSYUgDegkZKQQwUMIU4MA5AAUOABzYJyjfKgC26gK5FiAXlxGwAXSQA%2BJADoA1mHJLuEqcP2QIVAHMwZT1DExIkVAAjACt5ZHUNLRhIJCpsFFRkKigoVAB3EgBCDykvYQBfTwFRaVl4xSQlEEYo1GJyazoACyoIAAo6MBpEvxhUOEZouMJHOCpfM2YAFRhfLmsASjp9KgR8bsh%2Bqdj42wc6RG70A5goYn6AFk3Njy9fCH8g5VX1xxlkPpOGAkAhenAkI8aEgStUKoJrCBgIJBNpNKgAGZIDEQsx4poYoxwQjjOBKJAAMgpdQQvzAqBM-X6myQZgctSqoLW9JMmwA3IJCcSEKSkIl4LSYOoBizxEIudLHBjOABRKgHJki9S8GAstlIPRwCCoOSOPIBfoAA1W6iQABJRDAKnRLUgANQKllUpB0zD-MAIZn8ry%2B-2BzaCeECUMA5kC6Pcv2x4MCQTDBbqOSMZaONiYWxmMRMADK-wQqHUmmIXSQpcwRLg8ACSAq8dA3g7kgAfj3e32u-x5Z3h1JAFWESEA6uSAeD%2BkLt9t1InAoORIqdCEhXCv4Aa%2BiCYAglBADQZjKZB0i0zQM1mc6hnM4C0X5r4APLOBh1htNlt8pDAYBIIAoORIIcGQwEeORQJEzi8Ggs57AcKICIaECAqy9r9HKwiMHWYBUB0Na1COxF-gBwGgXuR56Ow8TLvB84RBMdHTPEG5uEg-ShKeJARkOaQkhMEAMLUEwqogbC8KJND7i21SSLgVBGvupJCUWqHpMgVReFp8qMEoaAIG%2BSg1gZb4AGo5FYvDEf%2BwjkWwlHHjRhB0XOByMUuK4seum4cVx4TELxwjPmAb7CdICBGBiWI6RU-JAA&eslintrc=&tsconfig=&tokens=false)

**Reply by fisker** [reply]
`Property[computed=true] > .key` is even shorter, and it should work.
